runtime/bindings/python/hal.cc

```diff
@@ -906,16 +906,19 @@ std::vector<std::string> HalDriver::Query() {
 }
 
 py::object HalDriver::Create(const std::string& device_uri,
-                             py::dict& driver_cache) {
+                             py::dict& driver_cache,
+                             std::optional<bool> clean) {
```

Comment:
It looks like you aren't actually cleaning anything.

How about changing this signature to just take an `std::optional<py::dict*> driver_cache`. If provided, that will be used/updated. If not, then no caching behavior?

Reply:
The cleaning is making a "clean" new driver object, which replaces whatever was in the cache.  I guess that wasn't clear enough, and hence is a failure of the comprehensibility of my API.  :-(

The intention, however best to express it, is that I didn't want to destroy all of the cached drivers, or require the driver to be separately managed.

Reply:
Ah. I guess I'm just being paranoid/protective about shared global state, and I did misread the intent and the comment (about making a request to clean implying that you wanted the cache cleaned).

I expect that use cases are going to break down between people who don't want to think about it using process global defaults and people who are more sophisticated just wanting to manage their own destiny without process global state making the resource timeline harder to understand. That's why I was thinking to keep the cached and uncached worlds separated: one uses global state and the other does not... Vs half ignoring and mutating the global state.

Reply:
This discussion is made obsolete by the change to the 3-function API.